### PR TITLE
feat: add xgrammar structured outputs and tool calling

### PR DESCRIPTION
**Pull request by ojus1**
## Summary
- refactor tinygrad LLM generation to expose logits before constrained token selection
- add optional xgrammar integration for GGUF tokenizers and OpenAI-compatible constrained decoding
- support structured outputs, required/forced tool choice, parallel tool calls, and streaming tool-call deltas
- add Qwen3.5 2B alias plus unit/null/external tests and an implementation plan doc

## Testing
- python -m pytest test/unit/test_llm_server.py test/null/test_llm_server.py test/null/test_llm_tokenizer.py test/unit/test_llm_xgrammar.py test/null/test_llm_xgrammar.py test/external/external_test_llm_xgrammar.py -q
- live-validated on macOS Metal with Qwen3.5:2b for:
  - JSON schema structured outputs
  - required single tool call
  - parallel tool calls
  - streaming tool calls

## Notes
- xgrammar remains an optional dependency path, added via extras/testing support
- real Qwen3.5:2b requests were validated against the local tinygrad server on Metal

### CODE REVIEW

**Comment by ojus1** [comment]
Closing this upstream PR per request; keeping changes only on the fork.
